File: parent-app/lib/core/security/crypto_helper.dart
import 'dart:convert';
import 'package:cryptography/cryptography.dart';

class CryptoHelper {
  static const String prefix = 'enc:v1:';
  static String familyPassphrase = 'ParentSecretPassphrase2026';
  static final _salt = utf8.encode('ParentalControlSalt2026');
  static SecretKey? _cachedKey;

  static final _pbkdf2 = Pbkdf2(
    macAlgorithm: Hmac.sha256(),
    iterations: 100000,
    bits: 256,
  );

  static final _aesGcm = AesGcm.with256bits();

  static void setFamilyPassphrase(String passphrase) {
    if (passphrase.trim().isNotEmpty && passphrase != familyPassphrase) {
      familyPassphrase = passphrase.trim();
      _cachedKey = null;
    }
  }

  static Future<SecretKey> _getOrCreateKey() async {
    if (_cachedKey != null) return _cachedKey!;
    final secretKey = await _pbkdf2.deriveKey(
      secretKey: SecretKey(utf8.encode(familyPassphrase)),
      nonce: _salt,
    );
    _cachedKey = secretKey;
    return secretKey;
  }

  /// Encrypt plaintext to `enc:v1:<iv_b64>:<cipher_b64>`
  static Future<String> encrypt(String plaintext) async {
    if (plaintext.trim().isEmpty) return plaintext;
    try {
      final key = await _getOrCreateKey();
      final plainBytes = utf8.encode(plaintext);
      final secretBox = await _aesGcm.encrypt(plainBytes, secretKey: key);

      final ivB64 = base64Encode(secretBox.nonce);
      // Combine ciphertext and 16-byte MAC tag to match Java/WebCrypto standard GCM output
      final combined = <int>[...secretBox.cipherText, ...secretBox.mac.bytes];
      final cipherB64 = base64Encode(combined);

      return '$prefix$ivB64:$cipherB64';
    } catch (_) {
      return plaintext;
    }
  }

  /// Decrypt `enc:v1:<iv_b64>:<cipher_b64>` back to plaintext
  static Future<String> decrypt(String text) async {
    if (!text.startsWith(prefix)) return text;
    try {
      final parts = text.substring(prefix.length).split(':');
      if (parts.length != 2) return text;

      final iv = base64Decode(parts[0]);
      final combined = base64Decode(parts[1]);
      if (combined.length < 16) return text;

      final cipherOnly = combined.sublist(0, combined.length - 16);
      final macBytes = combined.sublist(combined.length - 16);

      final key = await _getOrCreateKey();
      final secretBox = SecretBox(
        cipherOnly,
        nonce: iv,
        mac: Mac(macBytes),
      );

      final decryptedBytes = await _aesGcm.decrypt(secretBox, secretKey: key);
      return utf8.decode(decryptedBytes);
    } catch (_) {
      return text;
    }
  }
}

File: kids-agent/app/src/main/java/com/parentalcontrol/kidsagent/security/CryptoHelper.kt
package com.parentalcontrol.kidsagent.security

import android.util.Base64
import android.util.Log
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

object CryptoHelper {

    private const val TAG = "CryptoHelper"
    private const val ALGORITHM = "AES/GCM/NoPadding"
    private const val GCM_TAG_LENGTH = 128
    private const val IV_LENGTH = 12
    private const val ITERATIONS = 100000
    private const val KEY_LENGTH = 256
    private val SALT = "ParentalControlSalt2026".toByteArray(Charsets.UTF_8)
    private const val PREFIX = "enc:v1:"

    // Default family passphrase (matches parent dashboard default, can be overridden)
    var familyPassphrase = "ParentSecretPassphrase2026"
    private var cachedKey: SecretKey? = null

    @Synchronized
    private fun getOrCreateKey(): SecretKey {
        if (cachedKey != null) return cachedKey!!
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val spec = PBEKeySpec(familyPassphrase.toCharArray(), SALT, ITERATIONS, KEY_LENGTH)
        val tmp = factory.generateSecret(spec)
        val key = SecretKeySpec(tmp.encoded, "AES")
        cachedKey = key
        return key
    }

    @Synchronized
    fun updatePassphrase(newPassphrase: String) {
        if (newPassphrase.isNotBlank() && newPassphrase != familyPassphrase) {
            familyPassphrase = newPassphrase
            cachedKey = null
            Log.d(TAG, "E2EE family passphrase updated")
        }
    }

    fun encrypt(plaintext: String): String {
        if (plaintext.isBlank()) return plaintext
        return try {
            val key = getOrCreateKey()
            val cipher = Cipher.getInstance(ALGORITHM)
            val iv = ByteArray(IV_LENGTH).apply { SecureRandom().nextBytes(this) }
            val spec = GCMParameterSpec(GCM_TAG_LENGTH, iv)
            cipher.init(Cipher.ENCRYPT_MODE, key, spec)

            val ciphertext = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))
            val ivB64 = Base64.encodeToString(iv, Base64.NO_WRAP)
            val cipherB64 = Base64.encodeToString(ciphertext, Base64.NO_WRAP)

            "$PREFIX$ivB64:$cipherB64"
        } catch (e: Exception) {
            Log.e(TAG, "Encryption failed: ${e.message}")
            plaintext // Fallback to plaintext if error
        }
    }

    fun decrypt(encryptedText: String): String {
        if (!encryptedText.startsWith(PREFIX)) return encryptedText
        return try {
            val parts = encryptedText.removePrefix(PREFIX).split(":")
            if (parts.size != 2) return encryptedText

            val iv = Base64.decode(parts[0], Base64.NO_WRAP)
            val ciphertext = Base64.decode(parts[1], Base64.NO_WRAP)

            val key = getOrCreateKey()
            val cipher = Cipher.getInstance(ALGORITHM)
            val spec = GCMParameterSpec(GCM_TAG_LENGTH, iv)
            cipher.init(Cipher.DECRYPT_MODE, key, spec)

            val plainBytes = cipher.doFinal(ciphertext)
            String(plainBytes, Charsets.UTF_8)
        } catch (e: Exception) {
            Log.e(TAG, "Decryption failed: ${e.message}")
            encryptedText
        }
    }
}
